### PR TITLE
Integrate ANDI-10609

### PR DESCRIPTION
**Pull request by aamereller**

### CODE REVIEW

File: src/app.cpp [hunk 1]
```diff
@@ -53,6 +62,30 @@ bool LoadNpcapDlls()
 }
 #endif
 
+uint32_t create_can_error_frame(uint8_t* data, uint16_t flags, bool is_canfd) {
+	uint32_t can_id = 0x20000000;
+	bool overflow = (flags & TMP_OVERFLOW) != 0;
```
Comment (kayoub5) [inline src/app.cpp:67]: OVERFLOW means device dropped packets, it's not an error frame

File: src/app.cpp [hunk 2]
```diff
@@ -53,6 +62,30 @@ bool LoadNpcapDlls()
...
+	data[3] |= (flags & (TMP_CRC_DEL_ERROR << bitshift)) != 0 ? 0x18 : 0;
+	data[3] |= (flags & (TMP_ACK_DEL_ERRoR << bitshift)) != 0 ? 0x1B : 0;
+	data[3] |= (flags & (TMP_EOF_ERROR << bitshift)) != 0 ? 0x1A : 0;
+	if ((data[2] != 0) || (data[3] != 0)) {
+		can_id |= 0x00000008;
+	}
+	return can_id;
+	{
+		/* code */
+	}
```
Comment (kayoub5) [inline src/app.cpp:85]: What's this empty block?

File: src/app.cpp [hunk 3]
```diff
@@ -94,14 +127,20 @@ void transform(
...
 			can.can_id = ntoh32(*((uint32_t*)data));
-			can.len = data[4];
 			if (header.data_type == TECMP_DATA_CANFD)
 			{
 				can.flags |= CANFD_FDF;
 				can.flags |= header.data_flags & TMP_BITRATE_SWITCH ? CANFD_BRS : 0;
 				can.flags |= header.data_flags & TMP_ERROR_NODE_ACTIVE ? CANFD_ESI : 0;
 			}
-			memcpy(can.data, data + 5, can.len);
+			if ((header.data_flags & TMP_ERROR_MESSAGE) != 0) {
```
Comment (kayoub5) [inline src/app.cpp:136]: Why check only the TMP_ERROR_MESSAGE flag?

**Comment by kayoub5** [comment]
@aamereller please add unit tests

**Comment by RayanMRABET** [comment]
@kayoub5  The conversion to CAN-FD error frame is incorrect. Am I mistaken?

image

**Comment by kayoub5** [comment]
> @kayoub5 The conversion to CAN-FD error frame is incorrect. Am I mistaken?
> 

You are not mistaken

**Comment by kayoub5** [comment]
Done in #37, closing

**Pull request closed without merging**